sw/device/lib/dif/dif_rv_timer.c

```diff
@@ -0,0 +1,380 @@
+        timer->base_addr,
+        reg_for_hart(hart_id, RV_TIMER_TIMER_V_UPPER0_REG_OFFSET));
+
+    if (upper == overflow_check) {
+      *out = (((uint64_t)upper) << 32) | lower;
+      return kDifRvTimerOk;
+    }
+  }
+}
+
+dif_rv_timer_result_t dif_rv_timer_arm(const dif_rv_timer_t *timer,
+                                       uint32_t hart_id, uint32_t comp_id,
+                                       uint64_t threshold) {
+  if (timer == NULL || hart_id >= timer->config.hart_count ||
+      comp_id >= timer->config.comparator_count) {
+    return kDifRvTimerBadArg;
+  }
+
+  uint32_t lower = threshold;
+  uint32_t upper = threshold >> 32;
+
+  ptrdiff_t lower_reg =
+      reg_for_hart(hart_id, RV_TIMER_COMPARE_LOWER0_0_REG_OFFSET) +
+      (sizeof(uint64_t) * comp_id);
+  ptrdiff_t upper_reg =
+      reg_for_hart(hart_id, RV_TIMER_COMPARE_UPPER0_0_REG_OFFSET) +
+      (sizeof(uint64_t) * comp_id);
+
+  // First, set the upper register to the largest value possible without setting
+  // off the alarm; this way, we can set the lower register without setting
+  // off the alarm.
+  mmio_region_write32(timer->base_addr, upper_reg, UINT32_MAX);
+
+  // This can't set off the alarm because of the value we set above.
+  mmio_region_write32(timer->base_addr, lower_reg, lower);
+  // Finish writing the new value; this may set off an alarm immediately.
+  mmio_region_write32(timer->base_addr, upper_reg, upper);
+
+  return kDifRvTimerOk;
+}
+
+/**
+ * The number of comparators that the IP instantiation this library is compiled
+ * against has. In other words, this tells us how far the statically known IRQ
+ * register constants are from the start of the comparators, which influences
+ * the computation in `irq_reg_for_hart()`.
+ */
+static const ptrdiff_t kComparatorsInReggenHeader =
+    (RV_TIMER_INTR_ENABLE0_REG_OFFSET - RV_TIMER_COMPARE_LOWER0_0_REG_OFFSET) /
+    sizeof(uint64_t);
+
+/**
+ * Computes the appropriate register for a particular interrupt register, given
+ * a number of comparators.
+ *
+ * Currently, all IRQ registers are placed after the comparator registers,
+ * so the register offsets given by HW are not useable. The offsets need to be
+ * compensated for by a factor of `kComparatorsInReggenHeader` double-words..
+ *
+ * We also do not handle the case when comparator_count > 32, which would cause
+ * multiple interrupt registers to be generated.
+ */
+static ptrdiff_t irq_reg_for_hart(uint32_t hart_id, uint32_t comparators,
+                                  ptrdiff_t reg_offset) {
+  // Note that it is completely valid for this value to be negative: if this
+  // library is built for hardware with a large number of compartors, this value
+  // is necessarially negative when used with hardware with a small number of
+  // comparators.
+  ptrdiff_t extra_comparator_offset =
+      sizeof(uint64_t) * (comparators - kComparatorsInReggenHeader);
+  return reg_for_hart(hart_id, reg_offset) + extra_comparator_offset;
+}
```

Comment:
Yes, this should work. The comment for the config struct is clear, as long as its fields are set correctly for the `rv_timer` instance being used, this eliminates the need for recompilation.